ccbuilder/utils/utils.py

```diff
@@ -12,12 +12,18 @@
 
 
 def run_cmd(
-    cmd: str, capture_output: bool = False, additional_env: dict[str, str] = {}
+    cmd: str,
+    capture_output: bool = False,
+    additional_env: dict[str, str] = {},
+    print_output: bool = False,
 ) -> str:
     env = os.environ.copy()
     env.update(additional_env)
     res = subprocess.run(
-        shlex.split(cmd), capture_output=capture_output, check=True, env=env
+        shlex.split(cmd),
+        capture_output=(capture_output or not print_output),
```

Comment:
I'm a bit confused, why are there two booleans that control the same thing?

Also, won't `(capture_output or not print_output)` evaluate to `True` with the current default values?

Reply:
Because when you capture the output, it is not printed. Also, you can't print when you want to capture something without losing the timing for `stdout` and `stderr`. Iirc, this was an issue/complaint at one point.

It does evaluate to `True`, hence the output is not printed, as it is captured, but also not returned, as `capture_output = False`.

Maybe naming the parameter `do_not_print`/`do_not_print_when_not_capturing`  might be better?

Reply:
I think it would make sense to document this function as the parameters are a bit confusing. 

> losing the timing for `stdout` and `stderr`
There's anyway no guarantee for these, neither at the process level (unless it's flushing it's buffers after every print), nor at the build system level, so this is not an issue in the context of `ccbuidler`. 

My first thought was to completely remove these booleans and always capture and return the output. And the caller can decide if it should be printed. The downside is that for long running processes (e.g., building a compiler) you only get the output at the very end which is not nice.

Reply:
>Maybe naming the parameter `do_not_print`/`do_not_print_when_not_capturing` might be better?

The negative makes more sense because with the current code one may pass `print_output = True` and not see anything printed. 

Also, if nothing is captured or printed shouldn't it be instead of `capture_output=True` shouldn't the outputs be redirected to `DEVNULL`

Reply:
Here's an idea (better names are probably needed):
```
@dataclass(kw_only=True, frozen=True)
class ProcessResult:
    return_value: int
    stdout: str | None
    stderr: str | None

def run_cmd(
    cmd: str,
    additional_env: dict[str, str] = {},
    print_stdout_and_stderr_instead_of_capturing: bool = False,
) -> ProcessResult:
```
By default `ProcessResult.{stdout,stderr}` contain the captured values, if `print_stdout_and_stderr_instead_of_capturing` is `True` then they are instead printed normally.

Reply:
Instead of `print_stdout_and_stderr_instead_of_capturing: bool` maybe a `class OutputMode(Enum)` would make more sense?

Reply:
I fear we might end up with a re-implementation of  `subprocess.run`. 

Another alternative is to enable capturing by default. Commands that don't need their results to be captured don't lose anything and don't print anymore.

Reply:
>  But if ccbuilder is invoked from the command line it makes no sense to just dump all the output at the end.

Long running commands use `run_cmd_to_logfile` where you can monitor the results in the the specified log file or (as it is currently), the CLI. Changing `run_cmd` doesn't change that.

The newly suggested version works with file descriptors, so we are in `run_cmd_to_logfile` territory. I'd replace it with  `run_cmd` when it gets the file descriptors/`TextIO`.